### PR TITLE
Introduce EXT-X-SKIP support

### PR DESCRIPTION
**Pull request by theRealRobG**
### Description

This PR introduces support for the EXT-X-SKIP tag.

The important addition as a result of supporting this tag (outside of what is already possible by users of the library creating a custom tag descriptor to support it) is that we now properly update the `mediaSequence` property when calculating each `MediaSegmentTagGroup`. Before this change Mamba would provide incorrect values for each `mediaSequence` occurring after an `EXT-X-SKIP` tag (which could break users of the library that receive a playlist delta update and rely on the library's media sequence calculations).

Support for other playlist tags related the playlist delta updates (i.e. EXT-X-SERVER-CONTROL) will be done separately (in an attempt to keep this PR focused).

### Change Notes

* Added new value (`EXT_X_SKIP`) to the `PantosTag` enum and made the necessary updates for the `HLSTagDescriptor` conformance.
* Added new values to `PantosValue` to account for attributes related to EXT-X-SKIP (`.skippedSegments` and `.recentlyRemovedDateranges`).
* Updated `generateMediaGroups` for `HLSPlaylistStructure` to account for any `SKIPPED-SEGMENTS` when calculating media sequence values for segment tag groups.

### Pre-submission Checklist

- [x] I ran the unit tests locally before checking in.
- [x] I made sure there were no compiler warnings before checking in.
- [x] I have written useful documentation for all public code.
- [x] I have written unit tests for this new feature.

### CODE REVIEW

File: mambaSharedFramework/HLS Models/Playlist Structure/HLSPlaylistStructure.swift [hunk 1]
```diff
@@ -372,16 +372,44 @@ fileprivate struct HLSPlaylistStructureConstructor {
...
+            {
                 currentMediaSequence = startMediaSequence
             }
         }
-        
+
+        // account for any skip tags (since a delta update replaces all segments earlier than the skip boundary, the
+        // SKIPPED-SEGMENTS value will effectively update the current media sequence value of the first segment, so safe
+        // to do this here and not within the looping through media group tags below).
+        if skipTagIndices.count > 0 {
```
Comment (rmigneco) [inline mambaSharedFramework/HLS Models/Playlist Structure/HLSPlaylistStructure.swift]: This approach seems to be more consistent with how the rest of the project is structured but I do wonder how much we care if there are multiple skip tags in the playlist? It should never happen so I guess you would catch it in development?

Not something that needs too be addressed here but I guess the alternatives would be:
- just checking for most recent skip tag or
- failing the HLS parse/validation if there were multiple tags?

File: mambaSharedFramework/Pantos-Generic HLS Playlist Parsing/PantosValue.swift [hunk 2]
```diff
@@ -189,7 +189,22 @@ public enum PantosValue: String {
...
-    
+
+    /// Found in `.EXT_X_SKIP`.
+    ///
+    /// The value is a decimal-integer specifying the number of Media
+    /// Segments replaced by the EXT-X-SKIP tag.  This attribute is
+    /// REQUIRED.
+    case skippedSegments = "SKIPPED-SEGMENTS"
+
+    /// Found in `.EXT_X_SKIP`.
```
Comment (rmigneco) [inline mambaSharedFramework/Pantos-Generic HLS Playlist Parsing/PantosValue.swift:200]: thank you for the documentation

File: mambaTests/HLSPlaylistStructureAndEditingTests.swift [hunk 3]
```diff
@@ -937,3 +957,23 @@ let sample4SegmentPlaylist =
         "#EXTINF:2.002,\n" +
         "http://not.a.server.nowhere/segment4.ts\n" +
 "#EXT-X-ENDLIST\n"
+
```
Comment (rmigneco) [inline mambaTests/HLSPlaylistStructureAndEditingTests.swift:960]: For new tests, can we use triple-quote strings to make these mocks more readable, i.e.

```swift
let sampleDeltaUpdatePlaylist =
"""
#EXTM3U
#EXT-X-VERSION:9
#EXT-X-MEDIA-SEQUENCE:1
#EXT-X-SERVER-CONTROL:CAN-SKIP-UNTIL=12
#EXT-X-TARGETDURATION:2
#EXT-X-SKIP:SKIPPED-SEGMENTS=4
#EXTINF:2.002
http://not.a.server.nowhere/segment5.ts
#EXTINF:2.002
http://not.a.server.nowhere/segment6.ts
#EXTINF:2.002
http://not.a.server.nowhere/segment7.ts
#EXTINF:2.002
http://not.a.server.nowhere/segment8.ts
#EXTINF:2.002
http://not.a.server.nowhere/segment9.ts
#EXTINF:2.002
http://not.a.server.nowhere/segment10.ts
"""
```

Ran this through the test and it works

**Review by rmigneco — COMMENTED** [review]
I think this is good and mostly some minor comments.

Question:
Sometimes media sequence numbers are published along with EXTINF tags, i.e.
```
#EXTINF:2.000,842920575
```

In this case we trust the server that those MS numbers correctly account for skipped segments and require no updating, right?

File: mambaTests/HLSPlaylistStructureAndEditingTests.swift (hunk 3 above)
Comment (theRealRobG) [reply mambaTests/HLSPlaylistStructureAndEditingTests.swift:960]: Yes, that looks much more readable. Updated here: e6e31cafc326b3a5762d57515f25694814941361

File: mambaSharedFramework/HLS Models/Playlist Structure/HLSPlaylistStructure.swift (hunk 1 above)
Comment (theRealRobG) [reply mambaSharedFramework/HLS Models/Playlist Structure/HLSPlaylistStructure.swift]: Yes, I was trying to keep in line with what existed, but I think I see where you're coming from on this.

I've made an update to avoid unnecessary looping through all the tags in the playlist here: d7c4e21269820780d497c0fdb9992a31e5964cc6

What do you think about that? I split it off in its own commit so that it's easier to revert if you aren't comfortable with the change.

**Comment by theRealRobG** [comment]
> I think this is good and mostly some minor comments.
> 
> Question: Sometimes media sequence numbers are published along with EXTINF tags, i.e.
> 
> ```
> #EXTINF:2.000,842920575
> ```
> 
> In this case we trust the server that those MS numbers correctly account for skipped segments and require no updating, right?

I think that should be fine, because ultimately the application of the Playlist Delta Update is the responsibility of the server, and so if the server is publishing those media sequence numbers in the EXTINF titles, then it is also the responsibility of the server to keep those up to date after applying the EXT-X-SKIP. Anecdotally from when I made a POC for this, it should be fine for the server to manage this, as it's just filtering out information that it no longer needs to send to the client, but the information should all be already calculated.

Long story short, I think that shouldn't pose any issues for us here.

File: mambaSharedFramework/HLS Models/Playlist Structure/HLSPlaylistStructure.swift (hunk 1 above)
Comment (rmigneco) [reply mambaSharedFramework/HLS Models/Playlist Structure/HLSPlaylistStructure.swift]: I like it thanks.

**Review by rmigneco — APPROVED** [review]

**Comment by rmigneco** [comment]
👍 
Merge in `develop_1.x` and LGTM
